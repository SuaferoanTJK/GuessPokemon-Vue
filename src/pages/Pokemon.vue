<script>
import Picture from "@/components/Picture.vue";
import Options from "@/components/Options.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";
import { capitalize } from "@vue/shared";

export default {
  components: { Picture, Options },
  data() {
    return {
      pokemons: [],
      pokemon: null,
      showPokemon: false,
      showMessage: false,
      message: { text: null, valid: false },
      lifes: 3,
    };
  },
  methods: {
    async getPokemons() {
      this.pokemons = await getPokemonOptions();
      const randomInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemons[randomInt];
    },
    uppercase(text) {
      return capitalize(text);
    },
    checkAnswer(pokemonID) {
      this.showPokemon = true;
      if (pokemonID === this.pokemon.id)
        this.message = { text: "Well done!", valid: true };
      else {
        this.message = {
          text: `Incorrect, the pokemon is ${this.uppercase(
            this.pokemon.name
          )}`,
          valid: false,
        };
        this.lifes--;
      }

      this.showMessage = true;
    },
    resetData() {
      this.showPokemon = false;
      this.showMessage = false;
      this.pokemons = [];
      this.pokemon = null;
      this.getPokemons();
    },
    newGame() {
      this.resetData();
      this.lifes = 3;
    },
  },
  mounted() {
    this.getPokemons();
  },
};
</script>

<template>
  <!-- Lifes -->
  <div data-test="lifes" class="h-[10vh] px-4 flex items-center">
    <p class="text-lg font-semibold mr-2">Lifes:</p>
    <ul class="flex items-center">
      <li v-for="life in lifes" :key="life">
        <img src="@/assets/life.png" alt="life" class="w-10" />
      </li>
    </ul>
  </div>
  <div class="min-h-[90vh] flex flex-col justify-center">
    <div class="px-4 md:w-1/2 mx-auto">
      <!-- Title -->
      <div data-test="title" class="flex justify-center text-center">
        <h1 class="text-3xl font-bold text-secondary">
          Who is this
          <span class="flex items-center gap-2">
            <img src="@/assets/pokemon.png" alt="PokemonLogo" class="w-60" />
            <p class="text-7xl">?</p>
          </span>
        </h1>
      </div>
      <!-- Loader -->
      <div v-if="!pokemon" data-test="loader" class="mt-5 flex justify-center">
        <div class="lds-default">
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
          <div></div>
        </div>
      </div>
      <!-- Game -->
      <div v-else class="my-5">
        <Picture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
        <h2
          v-if="showMessage"
          class="text-center text-3xl font-semibold mt-4"
          :class="[message.valid ? 'text-green-700' : 'text-red-700']"
        >
          {{ message.text }}
        </h2>
        <Options
          :pokemons="pokemons"
          :showPokemon="showPokemon"
          @selection="checkAnswer($event)"
        />
        <div v-if="showPokemon" class="w-1/2 mx-auto">
          <button
            v-if="lifes > 0"
            @click="resetData"
            class="mt-10 rounded bg-secondary opacity-80 hover:opacity-100 text-white font-semibold px-4 py-3 w-full"
          >
            Next Pokemon
          </button>
          <button
            v-else
            @click="newGame"
            class="mt-10 rounded bg-red-600 opacity-80 hover:opacity-100 text-white font-semibold px-4 py-3 w-full"
          >
            New Game
          </button>
        </div>
      </div>
    </div>
  </div>
</template>
