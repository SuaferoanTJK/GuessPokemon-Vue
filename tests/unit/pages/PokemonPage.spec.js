import { shallowMount, mount } from "@vue/test-utils";
import { findByTestAttr } from "../utils";
import { pokemons } from "../utils";
import Pokemon from "@/pages/Pokemon";

describe("Pokemon Unit Testing", () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallowMount(Pokemon);
  });
  test("snapshot must be matched", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("Page components should be displayed", () => {
    const title = findByTestAttr(wrapper, "title").text();
    const lifes = findByTestAttr(wrapper, "lifes").findAll("li");
    const loader = findByTestAttr(wrapper, "loader").exists();
    expect(title).toBe("Who is this ?");
    expect(lifes.length).toBe(3);
    expect(loader).toBeTruthy();
  });
  test("getPokemons method called on mounted", () => {
    const getPokemonsSpy = jest.spyOn(Pokemon.methods, "getPokemons");
    shallowMount(Pokemon);
    expect(getPokemonsSpy).toHaveBeenCalled();
  });
  test("snapshot must be matched when pokemons are loaded", () => {
    const wrapper = mount(Pokemon, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showMessage: false,
          message: { text: null, valid: false },
          lifes: 3,
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("checkAnswer method", async () => {
    const wrapper = shallowMount(Pokemon, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showMessage: false,
          message: { text: null, valid: false },
          lifes: 3,
        };
      },
    });
    await wrapper.vm.checkAnswer(5);
    expect(wrapper.find("h2").exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBeTruthy();
    expect(wrapper.find("h2").text()).toBe(
      "Incorrect, the pokemon is Bulbasaur"
    );
    await wrapper.vm.checkAnswer(1);
    expect(wrapper.find("h2").text()).toBe("Well done!");
  });
});
