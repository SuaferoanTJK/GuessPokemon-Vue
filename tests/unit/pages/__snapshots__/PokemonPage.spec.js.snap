// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Unit Testing snapshot must be matched 1`] = `
<!-- Lifes -->
<div data-test="lifes" class="h-[10vh] px-4 flex items-center">
  <p class="text-lg font-semibold mr-2">Lifes:</p>
  <ul class="flex items-center">
    <li><img src="" alt="life" class="w-10"></li>
    <li><img src="" alt="life" class="w-10"></li>
    <li><img src="" alt="life" class="w-10"></li>
  </ul>
</div>
<div class="min-h-[90vh] flex flex-col justify-center">
  <div class="px-4 md:w-1/2 mx-auto">
    <!-- Title -->
    <div data-test="title" class="flex justify-center text-center">
      <h1 class="text-3xl font-bold text-secondary"> Who is this <span class="flex items-center gap-2"><img src="" alt="PokemonLogo" class="w-60"><p class="text-7xl">?</p></span></h1>
    </div><!-- Loader -->
    <div data-test="loader" class="mt-5 flex justify-center">
      <div class="lds-default">
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pokemon Unit Testing snapshot must be matched when pokemons are loaded 1`] = `
<!-- Lifes -->
<div data-test="lifes" class="h-[10vh] px-4 flex items-center">
  <p class="text-lg font-semibold mr-2">Lifes:</p>
  <ul class="flex items-center">
    <li><img src="" alt="life" class="w-10"></li>
    <li><img src="" alt="life" class="w-10"></li>
    <li><img src="" alt="life" class="w-10"></li>
  </ul>
</div>
<div class="min-h-[90vh] flex flex-col justify-center">
  <div class="px-4 md:w-1/2 mx-auto">
    <!-- Title -->
    <div data-test="title" class="flex justify-center text-center">
      <h1 class="text-3xl font-bold text-secondary"> Who is this <span class="flex items-center gap-2"><img src="" alt="PokemonLogo" class="w-60"><p class="text-7xl">?</p></span></h1>
    </div><!-- Loader -->
    <!-- Game -->
    <div class="my-5"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="Pokemon" class="mx-auto h-64 object-contain transition-all duration-700 brightness-0">
      <!--v-if-->
      <div class="mt-10 grid md:grid-cols-2 gap-4"><button class="rounded text-lg capitalize text-white hover:font-semibold bg-tertiary p-2 opacity-80 hover:opacity-100 cursor-pointer w-full shadow-md">bulbasaur</button><button class="rounded text-lg capitalize text-white hover:font-semibold bg-tertiary p-2 opacity-80 hover:opacity-100 cursor-pointer w-full shadow-md">ivysaur</button><button class="rounded text-lg capitalize text-white hover:font-semibold bg-tertiary p-2 opacity-80 hover:opacity-100 cursor-pointer w-full shadow-md">venusaur</button><button class="rounded text-lg capitalize text-white hover:font-semibold bg-tertiary p-2 opacity-80 hover:opacity-100 cursor-pointer w-full shadow-md">charmander</button></div>
      <!--v-if-->
    </div>
  </div>
</div>
`;
